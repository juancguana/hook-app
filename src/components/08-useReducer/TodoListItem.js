import React from 'react';

const TodoListItem = ({ todo, index, handleToggle, handleDelete }) => {
  return (
    <li key={todo.id} className='list-group-item'>
      <p
        className={`${todo.done && 'complete'}`}
        onClick={() => handleToggle(todo.id)}
      >{`${index + 1}. ${todo.desc}`}</p>
      <button className='btn btn-danger' onClick={() => handleDelete(todo.id)}>
        Borrar
      </button>
    </li>
  );
};

export default TodoListItem;
