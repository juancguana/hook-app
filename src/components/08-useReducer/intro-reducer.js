const initialState = [{ id: 1, todo: 'Comprar pan', done: false }];
const ADD = 'ADD';

const todoReducer = (state = initialState, action) => {
  if (action?.type === ADD) {
    return [...state, action.payload];
  }
  return state;
};

let todos = todoReducer();

const newTodo = { id: 2, todo: 'Comprar leche', done: false };

const addTodoAction = {
  type: ADD,
  payload: newTodo,
};

todos = todoReducer(todos, addTodoAction);
console.log(todos);
