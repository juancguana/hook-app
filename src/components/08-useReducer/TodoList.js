import React from 'react';
import { TodoListItem } from './TodoListItem';

export const TodoList = ({ todos, handleDelete, handleToggle }) => {
  return (
    <ul>
      {todos.map((todo, index) => (
        <TodoListItem
          todo={todo}
          key={todo.id}
          index={index}
          handleDelete={handleDelete}
          handleToggle={handleToggle}
        />
      ))}
    </ul>
  );
};
