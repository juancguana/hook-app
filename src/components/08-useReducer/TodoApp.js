import React, { useEffect, useReducer } from 'react';
import { TodoAdd } from './TodoAdd';
import { TodoList } from './TodoList';
import { todoReducer } from './todoReducer';
import './styles.css';

const init = () => {
  return JSON.parse(localStorage.getItem('todos')) || [];
};

export const TodoApp = () => {
  const [todos, dispatch] = useReducer(todoReducer, [], init);

  useEffect(() => {
    localStorage.setItem('todos', JSON.stringify(todos));
  }, [todos]);

  const handleDelete = (todoId) => {
    const action = {
      type: 'delete',
      payload: todoId,
    };
    dispatch(action);
  };

  const handleToggle = (todoId) => {
    dispatch({
      type: 'toggle',
      payload: todoId,
    });
  };

  const handleAddTodo = (newTodo) => {
    dispatch({
      type: 'add',
      payload: newTodo,
    });
  };

  return (
    <div>
      <h1>Todo App ({todos.length})</h1>
      <hr />
      <div className='row'>
        <div className='col-7'>
          <TodoList
            todos={todos}
            handleToggle={handleToggle}
            handleDelete={handleDelete}
          />
        </div>
        <div className='col-5'>
          <TodoAdd handleAddTodo={handleAddTodo} />
        </div>
      </div>
    </div>
  );
};
