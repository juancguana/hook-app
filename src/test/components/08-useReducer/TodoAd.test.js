import { shallow } from 'enzyme';
import { TodoAdd } from '../../../components/08-useReducer/TodoAdd';

describe('Test in <TodoAdd />', () => {
  const handleAddTodo = jest.fn();
  const wrapper = shallow(<TodoAdd handleAddTodo={handleAddTodo} />);

  test('should show correctly', () => {
    expect(wrapper).toMatchSnapshot();
  });

  test('should not call handleAddTodo', () => {
    const formSubmit = wrapper.find('form').prop('onSubmit');
    formSubmit({ preventDefault() {} });

    expect(handleAddTodo).toHaveBeenCalledTimes(0);
  });

  test('should call handleAddTodo', () => {
    const value = 'Aprender React';

    wrapper.find('input').simulate('change', {
      target: {
        value,
        name: 'description',
      },
    });

    const formSubmit = wrapper.find('form').prop('onSubmit');

    formSubmit({ preventDefault() {} });

    expect(handleAddTodo).toHaveBeenCalledTimes(1);

    expect(handleAddTodo).toHaveBeenCalledWith(expect.any(Object));
    expect(handleAddTodo).toHaveBeenCalledWith({
      id: expect.any(String),
      desc: value,
      done: false,
    });
    expect(wrapper.find('input').prop('value')).toBe('');
  });
});
