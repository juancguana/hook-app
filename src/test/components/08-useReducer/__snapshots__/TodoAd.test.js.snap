// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test in <TodoAdd /> should show correctly 1`] = `
<Fragment>
  <h4>
    Agregar TODO
  </h4>
  <hr />
  <form
    onSubmit={[Function]}
  >
    <input
      autoComplete="off"
      className="form-control"
      name="description"
      onChange={[Function]}
      placeholder="Aprender"
      type="text"
      value=""
    />
    <button
      className="btn btn-outline-primary mt-1 w-100"
      type="submit"
    >
      Agregar
    </button>
  </form>
</Fragment>
`;
